### PR TITLE
Feature/metrics

### PR DESCRIPTION
**Pull request by anvelezec**
This pull request has the implementation of mean absolute error, as well as some unit test cases.

### CODE REVIEW

**Comment by cgarciae** [comment]
@anvelezec please rebase :)

File: treex/metrics/mean_absolute_error.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+    Returns:
+        jnp.ndarray values needed to update Mean Absolute Error
+    """
+
+    y_true = y_true.astype(y_pred.dtype)
+    return jnp.mean(jnp.abs(y_pred - y_true), axis=-1)
+
+
+class MeanAbsolutError(Mean):
```
Comment (cgarciae) [inline treex/metrics/mean_absolute_error.py:24]: Typo: should be `MeanAbsoluteError`

File: tests/metrics/test_mean_absolute_error.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import jax.numpy as jnp
+import numpy as np
+
+import torch
+from torchmetrics import MeanAbsoluteError
+
+from treex.metrics.mean_absolute_error import MeanAbsolutError
```
Comment (cgarciae) [inline tests/metrics/test_mean_absolute_error.py:7]: When the typo is fixed these classes will have the same name, probably best to do something like:

```python
import torchmetrics as tm
import treex as tx
```
and later use them as:

```python
tm.MeanAbsoluteError(...)
tx.metrics.MeanAbsoluteError(...)
```

**Comment by cgarciae** [comment]
@anvelezec would be cool if we could do more exhaustive tests using hypothesis like the one in here:

https://github.com/cgarciae/treex/blob/3fa742166b12cb808aa33c1df7b0b159093c6209/tests/losses/crossentropy_test.py#L21

File: tests/metrics/test_mean_absolute_error.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        mean_absolute_error_instance = MeanAbsolutError()
+        mae_treex = mean_absolute_error_instance(**{"y_true": y_true, "y_pred": y_pred})
+
+        mean_absolute_error = MeanAbsoluteError()
+        mae_tm = mean_absolute_error(torch.from_numpy(y_pred), torch.from_numpy(y_true))
+        assert np.isclose(np.array(mae_treex), mae_tm.numpy())
+
+    def test_accumulative_mae(self):
+        mean_absolute_error_instance = MeanAbsolutError()
+        mean_absolute_error = MeanAbsoluteError()
```
Comment (cgarciae) [inline tests/metrics/test_mean_absolute_error.py:25]: Maybe differentiate these variable names by the framework e.g. `mae_tm` and `mae_tx`

File: treex/metrics/mean_absolute_error.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    Args:
+        y_pred: Predicted tensor
+        y_true: Ground truth tensor
+
+    Returns:
+        jnp.ndarray values needed to update Mean Absolute Error
+    """
+
+    y_true = y_true.astype(y_pred.dtype)
+    return jnp.mean(jnp.abs(y_pred - y_true), axis=-1)
```
Comment (cgarciae) [inline treex/metrics/mean_absolute_error.py:21]: Curious if `jnp.mean` here is actually necessary.

Comment (anvelezec) [reply treex/metrics/mean_absolute_error.py]: @cgarciae yes its not necessary sin it is inheriting from Mean.

File: pyproject.toml [hunk 5]
```diff
@@ -33,7 +33,9 @@ mkdocs = "^1.2.2"
 mkdocstrings = "^0.15.2"
 mkdocs-jupyter = { version = "^0.17.3", python = ">=3.7" }
 mkdocs-material = "^7.2.5"
-tensorflow-cpu = "^2.6.0"
+einops = "^0.3.2"
+torchmetrics = "^0.5.1"
+tensorflow = "^2.6.0"
```
Comment (cgarciae) [inline pyproject.toml:38]: Can que use `tensorflow-cpu` instead? Its much lighter for CI and we don't use it for training. For `elegy` I've been using:

```toml
tf-nightly-cpu = "2.8.0.dev20211103"
```

File: tests/metrics/test_mean_absolute_error.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+
+class TestMAE:
+    def test_mae_basic(self):
+
+        y_true = np.random.randn(8, 20, 20)
+        y_pred = np.random.randn(8, 20, 20)
+
+        mean_absolute_error_tx = tx.metrics.MeanAbsoluteError()
+        mae_treex = mean_absolute_error_tx(**{"y_true": y_true, "y_pred": y_pred})
```
Comment (cgarciae) [inline tests/metrics/test_mean_absolute_error.py:18]:
```suggestion
        mae_treex = mean_absolute_error_tx(y_true=y_true, y_pred=y_pred)
```

File: tests/metrics/test_mean_absolute_error.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        if use_sample_weight:
+            params.update({"sample_weight": sample_weight})
+        mae_treex = mean_absolute_error_tx(**params)
+
+        mean_absolute_error_tx = tx.metrics.MeanAbsoluteError()
+        if use_sample_weight:
+            y_true, y_pred = y_true[sample_weight == 1], y_pred[sample_weight == 1]
+        mae_treex_no_sample_weight = mean_absolute_error_tx(
+            **{"y_true": y_true, "y_pred": y_pred}
+        )
```
Comment (cgarciae) [inline tests/metrics/test_mean_absolute_error.py:52]:
```suggestion
        mae_treex_no_sample_weight = mean_absolute_error_tx(
            y_true=y_true, y_pred=y_pred
        )
```

File: tests/metrics/test_mean_absolute_error.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        y_true = np.random.randn(8, 20, 20)
+        y_pred = np.random.randn(8, 20, 20)
+
+        mean_absolute_error_tx = tx.metrics.MeanAbsoluteError()
+        mae_treex = mean_absolute_error_tx(**{"y_true": y_true, "y_pred": y_pred})
+
+        mean_absolute_error_tx_short = tx.metrics.MAE()
+        mae_treex_short = mean_absolute_error_tx_short(
+            **{"y_true": y_true, "y_pred": y_pred}
+        )
```
Comment (cgarciae) [inline tests/metrics/test_mean_absolute_error.py:102]:
```suggestion
        mae_treex_short = mean_absolute_error_tx_short(
            y_true=y_true, y_pred=y_pred
        )
```

File: treex/metrics/mean_absolute_error.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+                Optional weighting of each example. Defaults to 1. shape = `[batch_size, d0, .. dN]`
+        Returns:
+            Array with the cumulative mean absolute error.
+        """
+        values = _mean_absolute_error(y_pred, y_true)
+        return super().update(values, sample_weight)
+
+    def compute(self):
+        """Computes MAE based on inputs passed in to ``update`` previously."""
+        return super().compute()
```
Comment (cgarciae) [inline treex/metrics/mean_absolute_error.py:88]:
```suggestion
```

Comment (cgarciae) [reply treex/metrics/mean_absolute_error.py:88]: Not needed.

File: treex/metrics/reduce.py [hunk 10]
```diff
@@ -64,24 +64,23 @@ def update(
...
-            sample_weight = sample_weight
-
-            # Update dimensions of weights to match with values if possible.
-            # values, _, sample_weight = tf_losses_utils.squeeze_or_expand_dimensions(
-            #     values, sample_weight=sample_weight
-            # )
+            if sample_weight.ndim > values.ndim:
+                raise Exception(
+                    f"sample_weight dimention is higher than values, when masking values sample_weight dimention needs to be equal or lower than values dimension, currently values have shape equal to {values.shape}"
+                )
```
Comment (cgarciae) [inline treex/metrics/reduce.py:70]: Nice!

**Comment by cgarciae** [comment]
@anvelezec can you accept the suggested changes? This PR is good to go.

**Pull request merged**